<a href="https://colab.research.google.com/github/podyssea/InformationRetrieval/blob/main/InformationRetrievalExercise2_2210049p.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Information Retrieval Exercise 2 Notebook 


This is the template notebook for Exercise 2. The specification for the exercise and the corresponding Exercise 2 submission instance are available on the Moodle page of the course.

This exercise builds upon Exercise 1, and assumes you are now familiar with concepts we have introduced in both the introductory lab and Exercise 1, including:
 - [PyTerrier operators](https://pyterrier.readthedocs.io/en/latest/operators.html)
 - [Pyterrier apply transformers](https://pyterrier.readthedocs.io/en/latest/transformer.html)
 - [PyTerrier pt.Experiment()](https://pyterrier.readthedocs.io/en/latest/experiments.html)


## PyTerrier Setup

Firt, we need to install PyTerrier. If you have been using your own machine, you will need to upgrade PyTerrier to the latest version for running this Exercise 2 notebook.

In [ ]:
!date
!pip install python-terrier


Mon Mar  1 12:22:38 UTC 2021
     |████████████████████████████████| 2.8MB 4.0MB/s 
     |████████████████████████████████| 102kB 9.4MB/s 
     |████████████████████████████████| 122kB 40.7MB/s 
     |████████████████████████████████| 92kB 8.5MB/s 
     |████████████████████████████████| 92kB 7.6MB/s 
ERROR: datascience 0.10.6 has requirement folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
  Found existing installation: Sphinx 1.8.5
    Uninstalling Sphinx-1.8.5:
      Successfully uninstalled Sphinx-1.8.5


Start PyTerrier; This time we are specifying `tqdm='notebook'` to get pretty progress bars during retrieval. 

In [ ]:
import pyterrier as pt
if not pt.started():
  pt.init(tqdm='notebook')

## Index, Topics & Qrels for Exercise 2

You will need your login & password credentials from Exercise 1. We will be using again the "50pct" and the "trec-wt-2004" datasets from Exercise 1.


In [251]:
UNAME="TODO"
PWORD="TODO"

from pyterrier.datasets import STANDARD_TERRIER_INDEX_FILES, RemoteDataset

# we will again be using the "50pct" and "trec-wt-2004" datasets
Fiftypct =  pt.get_dataset("50pct",  user="2210049p", password="e54de6e4")
dotgov_topicsqrels = pt.get_dataset("trec-wt-2004")

On the other hand, you will be using a slightly updated index for Exercise 2. It is a bit bigger than the Exercise 1 index, hence it takes about 2-3 minutes to download to Colab.

In [252]:
!date
indexref = Fiftypct.get_index(variant="ex2")
index = pt.IndexFactory.of(indexref)
!date

Mon Mar  1 14:12:30 UTC 2021
14:12:30.914 [main] WARN  o.t.structures.CompressingMetaIndex - Structure meta reading data file directly from disk (SLOW) - try index.meta.data-source=fileinmem in the index properties file. 860.9 MiB of memory would be required.
Mon Mar  1 14:12:31 UTC 2021


Let's check out the new index. Compared to the index we used for Exercise 1, you can see that this index has `Field Names: [TITLE, ELSE]`, which means that we can provide statistics about how many times each term occurs in the title of each document (the "TITLE" field), vs the rest of the document (the "ELSE" field). Refer to Lecture 8 for more information about fields.

Let's also display the keys in the meta index - this is the metadata that we have stored for each document. You can see that we are storing the "url" and the "body" (content) of the document. These will particularly come in handy for Q2 and Q3 of Exercise 2, respectively.


In [253]:
print(index.getCollectionStatistics().toString())
print("In the meta index: " + str(index.getMetaIndex().getKeys()))

Number of documents: 807775
Number of terms: 2043788
Number of postings: 177737957
Number of fields: 2
Number of tokens: 572916194
Field names: [TITLE, ELSE]
Positions:   true

In the meta index: ['docno', 'url', 'title', 'body']


Finally, these are all of the topics and qrels (including the training and validation datasets) that you will need to conduct Exercise 2.

In [254]:
tr_topics = Fiftypct.get_topics("training")
va_topics = Fiftypct.get_topics("validation")

tr_qrels = Fiftypct.get_qrels("training")
va_qrels = Fiftypct.get_qrels("validation")

test_topics = dotgov_topicsqrels.get_topics("hp")
test_qrels = dotgov_topicsqrels.get_qrels("hp")

## Baseline Setup

We introduce here the BatchRetrieve for our baseline. Note that:
 - We are using PL2 as our weighting model to generate the sample (the candidate set of documents to re-rank).
 - We expose more document metadata, namely "url" and "body" for each document retrieved, which you will need to deploy your two new features. 
 - By setting `verbose=True`, we display a progress bar while retrieval executes.

In [255]:
firstpassUB = pt.BatchRetrieve(index, wmodel="PL2", metadata=["docno", "url", "body"], verbose=True)

Let's see the resulting output - you can see that there are now "url" and "body" attributed for each retrieved document. (We also display a progress bar, enabled by the `verbose=True`).

In [256]:
firstpassUB.search("chemical reactions")

qid   docid           docno  ... rank      score               query
0     1  513586  G18-38-1767991  ...    0  12.755546  chemical reactions
1     1   38544  G01-14-2537005  ...    1  11.906524  chemical reactions
2     1  707122  G26-06-3754605  ...    2  11.877550  chemical reactions
3     1  382754  G13-59-3981168  ...    3  11.858475  chemical reactions
4     1   70292  G02-16-2617043  ...    4  11.731490  chemical reactions
..   ..     ...             ...  ...  ...        ...                 ...
995   1  246965  G08-68-4141101  ...  995   6.290707  chemical reactions
996   1  611136  G22-04-3955177  ...  996   6.289822  chemical reactions
997   1  594957  G21-38-0191596  ...  997   6.287830  chemical reactions
998   1  280944  G09-85-3411646  ...  998   6.287759  chemical reactions
999   1  596471  G21-44-0134257  ...  999   6.285508  chemical reactions

[1000 rows x 8 columns]

# Standard list of features

Let's introduce the list of features we need to deploy a baseline learning-to-rank approach.

In [257]:
pagerankfile = indexref.toString().replace(".properties", "-pagerank.oos")
features = [
    "SAMPLE", #ie PL2
    "WMODEL:SingleFieldModel(BM25,0)",
    "QI:StaticFeature(OIS,%s)" % pagerankfile,
]

stdfeatures = pt.FeaturesBatchRetrieve(index, features, verbose=True)
stage12 = firstpassUB >> stdfeatures

This is our feature set. We will be using FeaturesBatchRetrieve to compute these extra features on the fly. Let's see the output. You can see that there is now a "features" column.

In [258]:
stage12.search("chemical reactions").head(2)

qid               query  ...           docno      score
0   1  chemical reactions  ...  G18-38-1767991  12.755546
1   1  chemical reactions  ...  G01-14-2537005  11.906524

[2 rows x 7 columns]

Let's look in more detail at the features. It is clear that there are 3 numbers for each document. The first is the PL2 score (1.27555456e+01 == 12.7555), the second is the BM25 score, and the third is the PageRank (a link analysis feature - discussed in more detail in Lecture 10)

In [259]:
stage12.search("chemical reactions").head(1).iloc[0]["features"]

array([1.27555456e+01, 3.09240788e+00, 1.05668333e-04])

# Q1

You now have everyting you need to attempt Q1. You will need to refer to the specification, and to PyTerrier's [learning to rank documentation](https://pyterrier.readthedocs.io/en/latest/ltr.html).

You should use a LightGBM LambdaMART implementation (not XGBoost), instantiated using the configuration suggested in the PyTerrier documentation.

Hints:
 - You will need to use the provided separate “training” and “validation” topic sets and qrels to train the learning-to-rank.
 - There is no need to vary the configuration of LightGBM from that in the documentation.

In [260]:
import lightgbm as lgb
lmart_l = lgb.LGBMRanker(task="train",
    min_data_in_leaf=1,
    min_sum_hessian_in_leaf=100,
    max_bin=255,
    num_leaves=7,
    objective="lambdarank",
    metric="ndcg",
    ndcg_eval_at=[1, 3, 5, 10],
    learning_rate= .1,
    importance_type="gain",
    num_iterations=10)

In [261]:
lmart_l_pipe = stage12 >> pt.ltr.apply_learned_model(lmart_l, form="ltr")

In [262]:
lmart_l_pipe.fit(tr_topics, tr_qrels, va_topics, va_qrels)

/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:118: UserWarning: Found `num_iterations` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))


[1]	valid_0's ndcg@1: 0.277778
[2]	valid_0's ndcg@1: 0.351852
[3]	valid_0's ndcg@1: 0.351852
[4]	valid_0's ndcg@1: 0.388889
[5]	valid_0's ndcg@1: 0.388889
[6]	valid_0's ndcg@1: 0.407407
[7]	valid_0's ndcg@1: 0.407407
[8]	valid_0's ndcg@1: 0.407407
[9]	valid_0's ndcg@1: 0.388889
[10]	valid_0's ndcg@1: 0.388889


In [263]:
pt.Experiment([stage12, lmart_l_pipe], test_topics, test_qrels, ["map"], names=["PL2 Baseline", "LTR (PL2 Sample)"], baseline = 0)

name       map  map +  map -  map p-value
0      PL2 Baseline  0.225110    NaN    NaN          NaN
1  LTR (PL2 Sample)  0.410703   45.0   15.0     0.000097

In [264]:
pt.Experiment([stage12, lmart_l_pipe], test_topics, test_qrels, ["P.5"], names=["PL2 Baseline", "LTR (PL2 Sample)"], baseline = 0)

name       P_5  P_5 +  P_5 -  P_5 p-value
0      PL2 Baseline  0.069333    NaN    NaN          NaN
1  LTR (PL2 Sample)  0.114667   22.0    6.0     0.001787

In [265]:
lmart_l_pipe.search("chemical reaction")

qid              query   docid  ...           docno     score  rank
367   1  chemical reaction  134364  ...  G04-47-3650428  0.142846     0
383   1  chemical reaction   53842  ...  G01-62-2712429  0.007221     1
1     1  chemical reaction   38544  ...  G01-14-2537005 -0.015507     2
23    1  chemical reaction  304750  ...  G10-73-3158821 -0.015507     3
26    1  chemical reaction  365058  ...  G12-95-3885770 -0.080003     4
..   ..                ...     ...  ...             ...       ...   ...
995   1  chemical reaction  246965  ...  G08-68-4141101 -0.375935   995
996   1  chemical reaction  611136  ...  G22-04-3955177 -0.375935   996
997   1  chemical reaction  594957  ...  G21-38-0191596 -0.375935   997
998   1  chemical reaction  280944  ...  G09-85-3411646 -0.375935   998
999   1  chemical reaction  596471  ...  G21-44-0134257 -0.375935   999

[1000 rows x 7 columns]

#Q2 - URL Length Features

In this block, please provide your code for Q2 concerning the URL length feature(s). There are different possible URL length features that you could implement (see specification) - you might try more than one implementation, and choose the one that is the most effective.

Some hints:

 - You will need to use a [pt.apply function](https://pyterrier.readthedocs.io/en/latest/apply.html) for computing your URL length feature(s). The dataframe of results obtained from the upstream transformer has all of the information you need.

 - You can use a `**` operator for combining feature sets.

 - Refer to the PyTerrier learning to rank documentation  concerning `features_importances_` for obtaining feature importances, as required in Q2(c).

 - You may wish to refer to Python's [`urlparse()`](https://docs.python.org/3/library/urllib.parse.html) function.

 - Use Python assertions to test that your feature implmentation(s) give the expected results. This is needed for Q2(b).


In [266]:
import numpy as np
def _url_length_feature(row):
    docid = row["docid"]
    content = index.getMetaIndex().getItem("url", docid)
    f1 = len(content)
    return np.array([f1])

lmart_l_urls = lgb.LGBMRanker(task="train",
    min_data_in_leaf=1,
    min_sum_hessian_in_leaf=100,
    max_bin=255,
    num_leaves=7,
    objective="lambdarank",
    metric="ndcg",
    ndcg_eval_at=[1, 3, 5, 10],
    learning_rate= .1,
    importance_type="gain",
    num_iterations=10)

In [267]:
secondpassUB = pt.BatchRetrieve(index, wmodel="PL2", metadata=["docno", "url", "body"], verbose=True)

In [268]:
url_length_feature = secondpassUB >> pt.apply.doc_features(_url_length_feature)

In [269]:
url_length_feature.search("chemical reaction")

qid   docid           docno  ...      score              query  features
0     1  513586  G18-38-1767991  ...  12.755546  chemical reaction      [50]
1     1   38544  G01-14-2537005  ...  11.906524  chemical reaction      [59]
2     1  707122  G26-06-3754605  ...  11.877550  chemical reaction      [62]
3     1  382754  G13-59-3981168  ...  11.858475  chemical reaction      [62]
4     1   70292  G02-16-2617043  ...  11.731490  chemical reaction      [43]
..   ..     ...             ...  ...        ...                ...       ...
995   1  246965  G08-68-4141101  ...   6.290707  chemical reaction      [39]
996   1  611136  G22-04-3955177  ...   6.289822  chemical reaction      [44]
997   1  594957  G21-38-0191596  ...   6.287830  chemical reaction      [55]
998   1  280944  G09-85-3411646  ...   6.287759  chemical reaction      [38]
999   1  596471  G21-44-0134257  ...   6.285508  chemical reaction      [47]

[1000 rows x 9 columns]

In [270]:
combined_features = secondpassUB >> (stage12 ** url_length_feature)

In [271]:
combined_features.search("chemical reaction")["features"][0]

array([1.27555456e+01, 3.09240788e+00, 1.05668333e-04, 5.00000000e+01])

In [272]:
boosted_model = combined_features >> pt.ltr.apply_learned_model(lmart_l_urls, form="ltr")

In [273]:
boosted_model.fit(tr_topics, tr_qrels, va_topics, va_qrels)

/usr/local/lib/python3.7/dist-packages/lightgbm/engine.py:118: UserWarning: Found `num_iterations` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))


[1]	valid_0's ndcg@1: 0.296296
[2]	valid_0's ndcg@1: 0.296296
[3]	valid_0's ndcg@1: 0.388889
[4]	valid_0's ndcg@1: 0.407407
[5]	valid_0's ndcg@1: 0.407407
[6]	valid_0's ndcg@1: 0.425926
[7]	valid_0's ndcg@1: 0.425926
[8]	valid_0's ndcg@1: 0.425926
[9]	valid_0's ndcg@1: 0.425926
[10]	valid_0's ndcg@1: 0.425926


In [274]:
pt.Experiment([lmart_l_pipe, boosted_model], test_topics, test_qrels, ["map"], names=["LTR (PL2 Sample)", "LTR (PL2 Sample) + URL Lengths"], baseline = 0)

name       map  map +  map -  map p-value
0                LTR (PL2 Sample)  0.410703    NaN    NaN          NaN
1  LTR (PL2 Sample) + URL Lengths  0.414101   31.0   15.0     0.920929

In [275]:
pt.Experiment([lmart_l_pipe, boosted_model], 
              test_topics, 
              test_qrels, 
              ["P.5"], 
              names=["PL2 Sample", "LTR (PL2 Sample) + URL Lengths"],
              baseline = 0)

name       P_5  P_5 +  P_5 -  P_5 p-value
0                      PL2 Sample  0.114667    NaN    NaN          NaN
1  LTR (PL2 Sample) + URL Lengths  0.125333    6.0    2.0     0.158688

In [283]:
feature_scores_0 = boosted_model >> pt.ltr.feature_to_score(0)
feature_scores_1 = boosted_model >> pt.ltr.feature_to_score(1)
feature_scores_2 = boosted_model >> pt.ltr.feature_to_score(2)
feature_scores_3 = boosted_model >> pt.ltr.feature_to_score(3)

In [284]:
feature_scores_0.search("chemical reaction")

qid   docid  ...                                           features rank
0     1  513586  ...  [12.755545561073266, 3.0924078763629836, 0.000...    0
1     1   38544  ...  [11.90652405775751, 10.789390732195702, 0.0002...    1
2     1  707122  ...  [11.877549705152717, 0.0, 2.3513691708978857e-...    2
3     1  382754  ...  [11.858474825809738, 0.0, 5.346185439867095e-0...    3
4     1   70292  ...  [11.731489696645252, 0.0, 3.778848211693258e-0...    4
..   ..     ...  ...                                                ...  ...
995   1  246965  ...  [6.290706693490568, 3.3517926096567896, 1.3237...  995
996   1  611136  ...  [6.28982231530779, 0.0, 5.268805694263197e-06,...  996
997   1  594957  ...  [6.287829625241223, 0.0, 4.7705552713888965e-0...  997
998   1  280944  ...  [6.287758720877289, 0.0, 3.6757891813812704e-0...  998
999   1  596471  ...  [6.285507649261742, 0.0, 1.3408854886104408e-0...  999

[1000 rows x 9 columns]

In [285]:
feature_scores_1.search("chemical reaction")

qid   docid  ...                                           features rank
1     1   38544  ...  [11.90652405775751, 10.789390732195702, 0.0002...    0
23    1  304750  ...  [10.754678131519476, 10.789390732195702, 0.000...    1
26    1  365058  ...  [10.672285487792138, 10.789390732195702, 0.0, ...    2
367   1  134364  ...  [7.601036625189637, 8.813689796380995, 0.00054...    3
746   1  291491  ...  [6.697042378409187, 7.107157539831338, 8.01493...    4
..   ..     ...  ...                                                ...  ...
994   1  330899  ...  [6.291750803824551, 0.0, 1.731820272381033e-05...  995
996   1  611136  ...  [6.28982231530779, 0.0, 5.268805694263197e-06,...  996
997   1  594957  ...  [6.287829625241223, 0.0, 4.7705552713888965e-0...  997
998   1  280944  ...  [6.287758720877289, 0.0, 3.6757891813812704e-0...  998
999   1  596471  ...  [6.285507649261742, 0.0, 1.3408854886104408e-0...  999

[1000 rows x 9 columns]

In [286]:
feature_scores_2.search("chemical reaction")

qid   docid  ...                                           features rank
976   1    4022  ...  [6.318325231594571, 3.6586746181002856, 0.0142...    0
383   1   53842  ...  [7.530192821910841, 4.722623239505648, 0.00559...    1
247   1  140313  ...  [8.112216889846962, 0.0, 0.0028162164132355645...    2
361   1  133170  ...  [7.634414969813151, 0.0, 0.002734609877893981,...    3
909   1  147692  ...  [6.428308910022694, 0.0, 0.00269154680969787, ...    4
..   ..     ...  ...                                                ...  ...
857   1  642605  ...   [6.48656012021121, 5.044170773226196, 0.0, 50.0]  995
878   1  101173  ...               [6.4565701650806275, 0.0, 0.0, 43.0]  996
883   1  606005  ...                [6.450549921826192, 0.0, 0.0, 39.0]  997
925   1  794948  ...                  [6.3924136678603, 0.0, 0.0, 47.0]  998
950   1  121441  ...                [6.362845360273893, 0.0, 0.0, 65.0]  999

[1000 rows x 9 columns]

In [287]:
feature_scores_3.search("chemical reaction")

qid   docid  ...                                           features rank
296   1  203365  ...              [7.8454483836056195, 0.0, 0.0, 142.0]    0
584   1  507171  ...  [6.999084800021656, 0.0, 8.244170604408633e-06...    1
153   1  483237  ...  [8.644131524133533, 0.0, 0.0001175168057951845...    2
164   1  483652  ...  [8.608031249702462, 0.0, 0.0001175168057951845...    3
766   1  458197  ...  [6.657190842315015, 0.0, 0.0001213773887388841...    4
..   ..     ...  ...                                                ...  ...
234   1  489641  ...  [8.181702644728457, 0.0, 6.180420943499062e-06...  995
799   1   78652  ...  [6.580948054790943, 2.509744155081957, 3.90305...  996
379   1  400160  ...  [7.549375935506184, 0.0, 1.641036613266748e-05...  997
207   1    4816  ...  [8.326352252220067, 0.0, 0.0007828056285719556...  998
976   1    4022  ...  [6.318325231594571, 3.6586746181002856, 0.0142...  999

[1000 rows x 9 columns]

#Q3 Proximity Search Feature

Now you will implement a new query-dependent feature, using the MinDist() function, as discussed in the specification. 

Hints:
 - Again, remember to use assertions to test your feature implementations. This will be needed for Q3(b)
 - Refer to the PyTerrier learning to rank documentation concerning features_importances_ for obtaining feature importances, as required in Q3(c)



In [324]:
def parser(row):
  docid = row["docid"]
  D = index.getMetaIndex().getItem("url", docid)

  return D

In [325]:
def _MinDist(a, b):
  
    if a == b :
       return np.array([0])
 
    # get individual words in a list
    words = parser.split(" ")
 
    # assume total length of the string as
    # minimum distance
    min_dist = len(words)+1
 
    # traverse through the entire string
    for index in range(len(words)):
 
        if words[index] == a:
            for search in range(len(words)):
 
                if words[search] == b: 
 
                    # the distance between the words is
                    # the index of the first word - the 
                    # current word index 
                    curr = abs(index - search) - 1;
 
                    # comparing current distance with 
                    # the previously assumed distance
                    if curr < min_dist:
                       min_dist = curr
 
    # w1 and w2 are same and adjacent
    return np.array([min_dist])


In [326]:
# D = "this is a query generated to test the function i have created"
# a = "test"
# b = "created"
# MinDist(a, b, D)

In [327]:
lmart_l_distance = lgb.LGBMRanker(task="train",
    min_data_in_leaf=1,
    min_sum_hessian_in_leaf=100,
    max_bin=255,
    num_leaves=7,
    objective="lambdarank",
    metric="ndcg",
    ndcg_eval_at=[1, 3, 5, 10],
    learning_rate= .1,
    importance_type="gain",
    num_iterations=10)

In [328]:
thirdpassUB = pt.BatchRetrieve(index, wmodel="PL2", metadata=["docno", "url", "body"], verbose=True)

In [329]:
term_distance_feature = thirdpassUB >> pt.apply.doc_features(_MinDist("Experiment", "Radiate"))

AttributeError: ignored

In [293]:
combined_features_with_dist = thirdpassUB >> (stage12 ** term_distance_feature)

In [294]:
boosted_model_dist = combined_features_with_dist >> pt.ltr.apply_learned_model(lmart_l_distance, form="ltr")

In [295]:
boosted_model_dist.fit(tr_topics, tr_qrels, va_topics, va_qrels)

TypeError: ignored

In [ ]:
pt.Experiment([lmart_l_pipe, boosted_model_dist], test_topics, test_qrels, ["map"], names=["LTR (PL2 Sample)", "LTR (PL2 Sample) + AggMinDist"], baseline = 0)

In [ ]:
pt.Experiment([lmart_l_pipe, boosted_model_dist], test_topics, test_qrels, ["P.5"], names=["LTR (PL2 Sample)", "LTR (PL2 Sample) + AggMinDist"], baseline = 0)

#Q4 A 5-feature Learning-to-Rank Model

You will now experiment with the LightGMB LambdaMART technique where you include both your added features (URL Length and AggMinDist) along the 3 initial features inc PL2 sample (5 features in total). 

You need to learn a *new* model when using your final selection of 5 features.

#Q5-Q6 Analysis & Reflection

The specification requests more analysis, including figures. Use more code here as needed to draw the graphs and prepare the data input needed for the table. For example, are there any patterns in the queries that are improved (or not) when adding your two deployed features? 

Once you have conducted your analysis, the specification asks you to reflect on what you have learnt from conducting the experiments and what your main conclusions are.

# That's all Folks

**Submission Instructions:** Complete this notebook. Please make your submission on the Moodle Exercise 2 submission instance. As part of your submission, you will need to hand-in the following:
- A **PDF report** (6 pages MAX) with your answers to questions Q1--Q6 of the specification
- Your **completed .ipynb Colab notbook** (use File... Download .ipynb).

Your PDF report along your ipynb notebook solution must be submitted by **Wednesday 24th March 2021, 4:30pm**.